File: csharp/BCUR/BCUR/FountainUtils.cs
namespace BlockchainCommons.BCUR;

/// <summary>
/// Utility functions for fountain encoding/decoding.
/// </summary>
internal static class FountainUtils
{
    /// <summary>
    /// Calculates the optimal fragment length given the data length and maximum fragment length.
    /// </summary>
    internal static int FragmentLength(int dataLength, int maxFragmentLength)
    {
        var fragmentCount = DivCeil(dataLength, maxFragmentLength);
        return DivCeil(dataLength, fragmentCount);
    }

    /// <summary>
    /// Integer division rounding towards positive infinity.
    /// </summary>
    internal static int DivCeil(int a, int b)
    {
        var d = a / b;
        var r = a % b;
        return r > 0 ? d + 1 : d;
    }

    /// <summary>
    /// Partitions data into fragments of the specified length, zero-padding the last fragment.
    /// </summary>
    internal static List<byte[]> Partition(byte[] data, int fragmentLength)
    {
        var paddingLen = (fragmentLength - (data.Length % fragmentLength)) % fragmentLength;
        var padded = new byte[data.Length + paddingLen];
        Array.Copy(data, padded, data.Length);

        var result = new List<byte[]>();
        for (int i = 0; i < padded.Length; i += fragmentLength)
        {
            var fragment = new byte[fragmentLength];
            Array.Copy(padded, i, fragment, 0, fragmentLength);
            result.Add(fragment);
        }
        return result;
    }

    /// <summary>
    /// Chooses which fragment indexes to combine for the given sequence number.
    /// For the first fragmentCount sequences, returns the single corresponding fragment.
    /// After that, uses Xoshiro256** to randomly select and combine fragments.
    /// </summary>
    internal static List<int> ChooseFragments(int sequence, int fragmentCount, uint checksum)
    {
        if (sequence <= fragmentCount)
        {
            return [sequence - 1];
        }

        var seqU32 = (uint)sequence;
        var seed = new byte[8];
        seed[0] = (byte)(seqU32 >> 24);
        seed[1] = (byte)(seqU32 >> 16);
        seed[2] = (byte)(seqU32 >> 8);
        seed[3] = (byte)seqU32;
        seed[4] = (byte)(checksum >> 24);
        seed[5] = (byte)(checksum >> 16);
        seed[6] = (byte)(checksum >> 8);
        seed[7] = (byte)checksum;

        var rng = Xoshiro256.FromBytes(seed);
        var degree = rng.ChooseDegree(fragmentCount);

        var indexes = new List<int>(fragmentCount);
        for (int i = 0; i < fragmentCount; i++)
        {
            indexes.Add(i);
        }
        var shuffled = rng.Shuffled(indexes);
        shuffled.RemoveRange(degree, shuffled.Count - degree);
        return shuffled;
    }

    /// <summary>
    /// XORs v2 into v1 in-place.
    /// </summary>
    internal static void Xor(byte[] v1, byte[] v2)
    {
        for (int i = 0; i < v1.Length; i++)
        {
            v1[i] ^= v2[i];
        }
    }
}

File: python/bc-ur/src/bc_ur/_fountain_utils.py
"""Utility functions for fountain encoding/decoding."""

from __future__ import annotations

from ._crc32 import crc32
from ._xoshiro256 import Xoshiro256


def _div_ceil(a: int, b: int) -> int:
    """Integer division rounding towards positive infinity."""
    d, r = divmod(a, b)
    return d + (1 if r > 0 else 0)


def fragment_length(data_length: int, max_fragment_length: int) -> int:
    """Calculate the effective fragment length for fountain encoding."""
    fragment_count = _div_ceil(data_length, max_fragment_length)
    return _div_ceil(data_length, fragment_count)


def partition(data: bytes, frag_length: int) -> list[bytes]:
    """Partition data into fragments, padding the last one with zeros."""
    padding_needed = (frag_length - (len(data) % frag_length)) % frag_length
    padded = data + b"\x00" * padding_needed
    return [padded[i : i + frag_length] for i in range(0, len(padded), frag_length)]


def choose_fragments(sequence: int, fragment_count: int, checksum: int) -> list[int]:
    """Choose which fragment indexes to combine for a given sequence number."""
    if sequence <= fragment_count:
        return [sequence - 1]

    seed = sequence.to_bytes(4, "big") + checksum.to_bytes(4, "big")
    xoshiro = Xoshiro256.from_data(seed)
    degree = xoshiro.choose_degree(fragment_count)
    indexes = list(range(fragment_count))
    shuffled = xoshiro.shuffled(indexes)
    return shuffled[:degree]


def xor_bytes(v1: bytearray, v2: bytes) -> None:
    """XOR v2 into v1 in-place. Both must have the same length."""
    if len(v1) != len(v2):
        raise ValueError(
            f"XOR operands must have equal length ({len(v1)} != {len(v2)})"
        )
    for i in range(len(v1)):
        v1[i] ^= v2[i]
